File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int checked[1000];
priority_queue<int, vector<int>, greater<int>> pq;

int solution(int coin, vector<int> cards) {
    // 초기 라운드 1
    int answer = 1;
    
    // 카드의 수
    int n = cards.size();
    
    // 초기 카드로 가능한 경우 체크
    for(int i = 0; i < n / 3; ++i){
        checked[cards[i]] = 1;
        
        if(checked[n + 1 - cards[i]] != 0){
            // 0의 비용으로 PASS
            pq.push(0);
        }
    }
    
    // 라운드 진행
    for(int i = n / 3; i < n; ++i){
        checked[cards[i]] = 2;
        
        if(checked[n + 1 - cards[i]] != 0){
            // checked 값에 따라서 1 or 2의 비용
            pq.push(checked[n + 1 - cards[i]]);
        }
        
        // 2장 뽑을 때 마다
        if(i % 2 != 0){
            // 라운드 진행 불가
            if(pq.empty() || coin < pq.top()){
                break;
            }
            // 다음 라운드 진행
            else{
                coin -= pq.top();
                pq.pop();
                
                answer++;
            }
        }
    }
    
//     while(!pq.empty() && pq.top() <= coin){
//         answer++;
//         coin -= pq.top();
//         pq.pop();
//     }
    
//     if(answer > n / 3 + 1) answer = n / 3 + 1;
    
    return answer;
}

File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.java
import java.util.*;

class Solution {
    public int solution(int coin, int[] cards) {
        int answer = 0;

        int n = cards.length;

        int[] needCoin = new int[n + 1];
        for(int i = 0; i < n + 1; ++i) needCoin[i] = -1;
        
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for(int i = 0; i < n / 3; ++i) {
            if(needCoin[n + 1 - cards[i]] == 0) {
                needCoin[n + 1 - cards[i]] = -1;
                pq.offer(0);
            }
            else {
                needCoin[cards[i]] = 0;
            }
        }
        
        for(int i = n / 3; i <= n; i += 2) {
            answer++;
            if(i == n) {
                break;
            }
            
            needCoin[cards[i]] = 1;
            if(needCoin[n + 1 - cards[i]] != -1) {
                pq.offer(needCoin[n + 1 - cards[i]] + needCoin[cards[i]]);
            }
            
            needCoin[cards[i + 1]] = 1;
            if(needCoin[n + 1 - cards[i + 1]] != -1) {
                pq.offer(needCoin[n + 1 - cards[i + 1]] + needCoin[cards[i + 1]]);
            }
            
            if(!pq.isEmpty() && pq.peek() <= coin) {
                coin -= pq.poll();
            }else{
                break;
            }
        }
        
        return answer;
    }
}